### ISSUE TITLE
Automate pruning Eth1 nodes taks 

### ISSUE THREAD

**Issue by pablomendezroyo** [dappnode/DNP_DAPPMANAGER#915]
# Feature request 

### Description
The task of pruning Geth (and any other Eth1 node) is commonly done by DAppNode users to decrease the disk space used.

There is 

### Proposal
Allow users to prune Eth1 nodes directly from the UI. This can be easily achieved with a host script. Ideally this task should be performed peridically so users do not have to worry about it. 

**Resources**
- Medium post [how to prune geth](https://medium.com/@FlyGuyInTheSky/how-to-prune-the-dappnode-geth-package-to-gain-more-disk-space-6fd980850046)
- [Proposal to automate tasks with ofelia](https://github.com/dappnode/DNP_DAPPMANAGER/issues/855)

**Comment by dapplion** [comment]
That would be really nice!

**Comment by alexpeterson91** [comment]
> That would be really nice! To prune, does the main geth process have to stop?

Yes. It needs to be stopped to add the runtime flag to enable pruning which when finished needs to be removed and replaced with normal default runtime flags.   so dependent packages such as Prysm Lighthouse and Teku need a fallback ETH1 service defined to keep running while pruning is running. From what I've gathered the process takes roughly 3-4 hours on a normal machine most of our users run. But the complication for pruning automation that I've found is that you need to be parsing the logs to wait for the output saying pruning complete at which point you need to restart the service with the pruning flag removed and the default, or actually whatever other settings the user may have added to the EXTRA_OPTS ENV field.   Possible definitely but likely have to Have a script running  combing through the logs as they run for hours looking for errors and exceptions thrown or for the completion notification in order for it to resume normal function after being manually run by the user or periodically by a scheduled service.

**Comment by dapplion** [comment]
@alexpeterson91 After a complete run of pruning, if you restart the process with the prune flag again: does pruning take 3-4 hours again? Or is it much faster?

**Comment by alexpeterson91** [comment]
AFAIK it cannot prune when it’s already pruned and it needs to have the flag for the pruning state removed and then be restarted to make geth sync again. It must be stopped for pruning. And restarted at the end of the process.

**Comment by dapplion** [comment]
Parsing logs is a very sketchy option, it usually breaks after enough time passes, is there any other way to detect it's done? Maybe we can ask Geth to add a feature where when done with pruning it creates a flag file

**Comment by alexpeterson91** [comment]
Again AFAIK the only way to know is via the logs because everything else is shut down the apis etc. not sure how you’d know when it’s completed and it needs to complete or it will be corrupted. There were other things I found more recently about needing at least 50gb of open storage for pruning to begin too.  I know some group has a script for ethdocker I think but I didn’t look into it so don’t know if it’s just a cron job with timers or it’s more complex. I’ll try and find it again. But I couldn’t find any other mentions of the ability to auto prune geth anywhere.

**Comment by alexpeterson91** [comment]
https://gist.github.com/yorickdowne/3323759b4cbf2022e191ab058a4276b2 

This was the only script I could find but it seems to just be that a script that needs human intervention to check logs. So not helpful.

**Comment by dapplion** [comment]
> https://gist.github.com/yorickdowne/3323759b4cbf2022e191ab058a4276b2 This was the only script I could find but it seems to just be that a script that needs human intervention to check logs. So not helpful.
> […](#)
> Sent from my iPhone
> On Feb 3, 2022, at 10:57 PM, Lion - dapplion ***@***.***> wrote: ﻿ Parsing logs is a very sketchy option, it usually breaks after enough time passes, is there any other way to detect it's done? Maybe we can ask Geth to add a feature where when done with pruning it creates a flag file — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.

Check https://github.com/ethereum/go-ethereum/issues/24344#issuecomment-1031096208 what do you think would be best short term?

**Comment by yorickdowne** [comment]
Interesting. I think RocketPool does it by stopping the service, touching a lock file, then starting the service. Which means that when it finishes, docker will automatically restart it. 

https://github.com/rocket-pool/smartnode/blob/482d908de3d79192b9ef10a260c5d85cea35f350/rocketpool-cli/service/service.go#L402

The entrypoint script just looks for that lock file, runs with prune and removes it if found, runs geth normally if not.

https://github.com/rocket-pool/smartnode-install/blob/master/amd64/rp-smartnode-install/network/mainnet/chains/eth1/start-node.sh

**Comment by alexpeterson91** [comment]
I completely forgot rocket pool was the only other solution for auto pruning I could find in the wild already. Thanks for bringing it back up I totally forgot to dig into their code for auto prune.

**Comment by pablomendezroyo** [comment]
@yorickdowne thanks for the reference!

So it seems like they run a "prune container" with the ethclient volume attached to it:
- The command https://github.com/rocket-pool/smartnode/blob/482d908de3d79192b9ef10a260c5d85cea35f350/shared/services/rocketpool/client.go#L614
- The prune container: https://github.com/rocket-pool/smartnode/blob/e69cb9eb91fe6784cad4f01810d182bbe4de7da8/docker/rocketpool-prune-provision
- The entrypoint of the prune container: https://github.com/rocket-pool/smartnode/blob/e69cb9eb91fe6784cad4f01810d182bbe4de7da8/rocketpool/prune_provision.sh

The prune container creates this `lock` file in the ethclient volume. Then the ethclient container is restarted and in its entrypoint is checked for the `lockfile`, so if it exists then it start prunning:
- file check https://github.com/rocket-pool/smartnode-install/blob/3d4f334e5ace27be931972b89465a62fc4146b9d/amd64/rp-smartnode-install/network/mainnet/chains/eth1/start-node.sh#L8
- wait for prunning to finish with `;`  https://github.com/rocket-pool/smartnode-install/blob/3d4f334e5ace27be931972b89465a62fc4146b9d/amd64/rp-smartnode-install/network/mainnet/chains/eth1/start-node.sh#L10


I like the approach, however, I dont like the idea of having to do tasks on both sides: dappmanager and ethclients. Ideally we want to do everything from the dappmanager

**Comment by yorickdowne** [comment]
Yeah that’s what they do. I’ve done the same thing now. I don’t really see a way around having some kind of entrypoint for Geth to make this possible. If your dappmanager is always active and able to check things periodically, you could also run prune in a container named Geth-prune and then restart the proper Geth when the prune container is no longer active, checked via docker ps or something like it.

whatever you do will be temporary, as Geth will Soon(tm) no longer need to be offline pruned.

**Comment by 3alpha** [comment]
First one seems most elegant. What do you think?

**Comment by pablomendezroyo** [comment]
@3alpha we need to retake this! Implement automate prunning for Execution Clients that allow to do it online

**Comment by alexpeterson91** [comment]
I think some things need auto pruning enabled by default but i definitely agree that it should be user configurable as @3alpha suggested

**Comment by pablomendezroyo** [comment]
@3alpha should we close this due to the upcoming geth feature for ancient data
